app/models/planned_disbursement.rb

```diff
@@ -17,4 +19,10 @@ class PlannedDisbursement < ApplicationRecord
     :financial_quarter,
     :financial_year
   validates :value, inclusion: {in: 0.01..99_999_999_999.00}
+
+  def only_one_original
+    if PlannedDisbursement.find_by(financial_quarter: financial_quarter, financial_year: financial_year, parent_activity: parent_activity, planned_disbursement_type: :original).present?
+      errors.add(:base, I18n.t("activerecord.errors.models.planned_disbursement.attributes.planned_disbursement_type.only_one_original", financial_quarter: financial_quarter, financial_year_start: financial_year, financial_year_end: financial_year + 1))
```

Comment:
I'm wondering if it's currently possible to implement this by putting a unique index on these columns, rather than implementing as a validation rule? As I suggested in https://gist.github.com/jcoglan/dcb4e87652fd172a83f9263973cad2da, we have a use case where we need to enter "revisions" which requires either creating a new `PlannedDisbursement` for the relevant activity and quarter, or amending the existing one. In this case the caller needs to check which of these it should be doing, or using `upsert` or `create_or_find_by` to produce the right result. Trying to create a new one, getting a validation error, and then updating the existing one is a more awkward way of implementing that and doesn't ensure the data is consistent.

I'm also confused about the `planned_disbursement_type: :original` condition -- shouldn't there be a single planned disbursment per activity/quarter _of any type_, with the first one in the history implicitly being the original?

Reply:
I was hesitent to apply a restriction on the DB as I know we have data in production that would not meet the condition (so my expectation is the migration would fail), in my head the plan is to introduce the changes to make forecasts work as expected, fix the data in production and then be in a postion to apply tighter DB level restrictions - happy to admit this may not be the best approach, but it is the one I settled on - I would love to talk through another approach if we think there is a better one?

This validation is used to prevent a user adding what they believe is a new forecast (by clicking the add new forecast button) and accidentally tyring to add it to a financial quarter that already has a forecast - I believe this is different to the user who is actively looking to revise an existing forecast (for which there is UI in the next piece of work https://trello.com/c/WB5Q9F9d)

Not sure I agree with your second statement, for a given activity and financial quarter I expect only one 'original' (the first) and then any number of revisions (a maximum of one per reporting quarter) for the life of the activity?

Happy to talk this through as a wider piece of work to try and address your concerns, but I do think it's important to see all the steps that we have planned?